bash-preexec.sh

```diff
@@ -239,6 +250,23 @@ __bp_preexec_invoke_exec() {
         return
     fi
 
+    # If we have removed "ignorespace" or "ignoreboth" from HISTCONTROL
+    # during setup, we need to remove commands that start with a space from
+    # the history ourselves.
+
+    # With bash 5.0 or above, we could have just ran
```

Comment:
We can switch the implementation based on `BASH_VERSINFO`. Reducing a fork for Bash 5.0 is still nice.

Reply:
While I agree with you partly, I have to say that keeping compatibility with ancient bash versions is also a problem. Why do people have to "suffer", because of it?
How many people are really on bash <5?

I come from a performance background and doing a regex where it is not needed seems a waste to me. Of course this is not a huge strain on resources, but preexec is a hook, which means it is called all the time, thus is should be as sleek and performant as possible, wouldn't you agree?

Maintenance is a good point, however, I would still favor the code path that is most likely taken, which in my book is the bash >=5 path.

Reply:
Thank you for updating. Bash-preexec already has a version check and supports Bash 3.1+. I'm not a maintainer, but I think it is not realistic to drop the support for Bash < 5. The situation for Bash and bash-preexec seems to be different from a typical project.

- Maybe we can drop the support for 3.1, but I think we still need to support 3.2 because 3.2 will continue to be shipped with macOS for the license issue.
- Also, LTS distributions will continue to provide a Bash version for ten years. Since Bash is tightly bound to the system, it's non-trivial to update the system-installed Bash for distributors. Although users can install the latest version of Bash in their home directory, it is difficult to force all the users to do so correctly.
- In a typical project, another option might be to maintain several branches of versions, where new features are added to the well-tested latest version, while older versions receive the corrections for real problems. However, the *feature set* of bash-preexec is stable and no feature was added for a long time, yet the implementation has been always unstable and incomplete. There hasn't been any *stable* version, and we would probably want to apply many patches retroactively even if we adopted such a model. Anyway, this model would add even more maintenance costs.

Although you seem to suggest two options, dropping support for Bash < 5.0 entirely or `if`-statement, if you care about the test coverage, I think we will need to keep the current version based on the old Bash features.

Reply:
I've updated my comment above. I just wanted to clarify that I didn't mean to drop support for bash < 5. I meant the `if` statement should be used to allow bash 5 users to use the more performant processing option.